radix-engine/src/track/interface.rs

```diff
@@ -102,24 +107,98 @@ pub trait SubstateStore {
         substate_key: &SubstateKey,
         flags: LockFlags,
         virtualize: F,
-    ) -> Result<(u32, bool), AcquireLockError>;
+    ) -> Result<(u32, StoreAccessInfo), AcquireLockError>;
 
     /// Releases a lock.
     ///
     /// # Panics
     /// - If the lock handle is invalid.
-    fn release_lock(&mut self, handle: u32);
+    fn release_lock(&mut self, handle: u32) -> StoreAccessInfo;
 
     /// Reads a substate of the given node partition.
     ///
     /// # Panics
     /// - If the lock handle is invalid
-    fn read_substate(&mut self, handle: u32) -> &IndexedScryptoValue;
+    fn read_substate(&mut self, handle: u32) -> (&IndexedScryptoValue, StoreAccessInfo);
 
     /// Updates a substate.
     ///
     /// # Panics
     /// - If the lock handle is invalid;
     /// - If the lock handle is not associated with WRITE permission
-    fn update_substate(&mut self, handle: u32, substate_value: IndexedScryptoValue);
+    fn update_substate(
+        &mut self,
+        handle: u32,
+        substate_value: IndexedScryptoValue,
+    ) -> StoreAccessInfo;
+}
+
+#[derive(Clone)]
+pub struct StoreAccessInfo(Vec<StoreAccess>);
+
+impl StoreAccessInfo {
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    pub fn with_vector(vector: Vec<StoreAccess>) -> Self {
+        Self(vector)
+    }
+
+    pub fn data(&self) -> &Vec<StoreAccess> {
+        &self.0
+    }
+
+    pub fn get_whole_size(&self) -> usize {
```

Comment:
Can we split into `total_read_size` and `total_write_size`? In the future, we may also need to relate costing to `number_of_keys` depending on our database characteristics.